File: Data Structures & Algorithms/balanced-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool result = true;
    bool isBalanced(TreeNode* root) {
        dfs(root);

        return result;    
    }

    int dfs(TreeNode* root) {
        if (!root) {
            return 0;
        }

        int left = dfs(root->left);
        int right = dfs(root->right);

        if (abs(right - left) > 1) {
            result = false;
        }

        return 1 + max(left, right);
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        good_nodes = 0
        queue = deque([(root, root.val)])

        while queue:
            node, path_val = queue.popleft()

            if node.val >= path_val:
                good_nodes += 1
            
            new_val = max(path_val, node.val)

            if node.left:
                queue.append((node.left, new_val))
            
            if node.right:
                queue.append((node.right, new_val))
        
        return good_nodes